packages/ui/VirtualList/VirtualListBase.js

```diff
@@ -23,728 +27,771 @@ const gridListItemSizeShape = PropTypes.shape({
-
-		/**
-		 * `true` if rtl, `false` if ltr.
-		 *
-		 * @type {Boolean}
-		 * @private
-		 */
-		rtl: PropTypes.bool,
-
-		/**
-		 * Spacing between items.
-		 *
-		 * @type {Number}
-		 * @default 0
-		 * @public
-		 */
-		spacing: PropTypes.number,
-
-		/**
-		 * Called to execute additional logic in a themed component when updating states and bounds.
-		 *
-		 * @type {Function}
-		 * @private
-		 */
-		updateStatesAndBounds: PropTypes.func
-	}
-
-	static defaultProps = {
-		cbScrollTo: nop,
-		data: [],
-		dataSize: 0,
-		direction: 'vertical',
-		overhang: 3,
-		pageScroll: false,
-		spacing: 0
-	}
-
-	constructor (props) {
-		super(props);
-
-		this.state = {firstIndex: 0, numOfItems: 0};
-	}
-
-	componentWillMount () {
-		if (this.props.clientSize) {
-			this.calculateMetrics(this.props);
-			this.updateStatesAndBounds(this.props);
+const VirtualListBaseFactory = (type) => {
+	return class VirtualListCore extends Component {
+		/* No displayName here. We set displayName to returned components of this factory function. */
+
+		static propTypes = /** @lends ui/VirtualList.VirtualListNative.prototype */ {
+			/**
+			 * The `render` function for an item of the list receives the following parameters:
+			 * - `data` is for accessing the supplied `data` property of the list.
+			 * > NOTE: In most cases, it is recommended to use data from redux store instead of using
+			 * is parameters due to performance optimizations.
+			 *
+			 * @param {Object} event
+			 * @param {Number} event.data-index It is required for Spotlight 5-way navigation. Pass to the root element in the component.
+			 * @param {Number} event.index The index number of the componet to render
+			 * @param {Number} event.key It MUST be passed as a prop to the root element in the component for DOM recycling.
+			 *
+			 * Data manipulation can be done in this function.
+			 *
+			 * > NOTE: The list does NOT always render a component whenever its render function is called
+			 * due to performance optimization.
+			 *
+			 * Usage:
+			 * ```
+			 * renderItem = ({index, ...rest}) => {
+			 *		delete rest.data;
+			 *
+			 *		return (
+			 *			<MyComponent index={index} {...rest} />
+			 *		);
+			 * }
+			 * ```
+			 *
+			 * @type {Function}
+			 * @public
+			 */
+			itemRenderer: PropTypes.func.isRequired,
+
+			/**
+			 * Size of an item for the list; valid values are either a number for `VirtualList`
+			 * or an object that has `minWidth` and `minHeight` for `VirtualGridList`.
+			 *
+			 * @type {Number|ui/VirtualList.gridListItemSizeShape}
+			 * @private
+			 */
+			itemSize: PropTypes.oneOfType([
+				PropTypes.number,
+				gridListItemSizeShape
+			]).isRequired,
+
+			/**
+			 * The render function for the items.
+			 *
+			 * @type {Function}
+			 * @private
+			 */
+			itemsRenderer: PropTypes.func.isRequired,
+
+			/**
+			 * Callback method of scrollTo.
+			 * Normally, [ScrollableNative]{@link ui/Scrollable.ScrollableNative} should set this value.
```

Comment:
Would you convert `ScrollableNative` to `Scrollable`?

Reply:
Updated.